File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int l = 0;
        int r = 0;
        int max_f = 0;
        int max_length = 0;
        int length = 0;
        std::unordered_map<char,int> freq_mp;
        while(r < s.size()){
            freq_mp[s[r]]++;
            max_f = max(max_f,freq_mp[s[r]]);
            length = r-l+1;
            while(length - max_f > k && l<=r){
                freq_mp[s[l]]--;
                max_f = 0;
                for(const auto& elem : freq_mp) max_f = max(max_f,elem.second);
                l++;
                length = r-l+1;
            }
            max_length = max(length,max_length);
            r++;
        }
        return max_length;
        
        
    }
};

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freq = {}
        left, max_freq, max_length = 0, 0, 0

        for right in range(len(s)):
            freq[s[right]] = freq.get(s[right],0)+1
            max_freq = max(max_freq,freq[s[right]])

            window_size = right - left + 1

            if window_size - max_freq > k:
                freq[s[left]] -=1
                left +=1
            
            max_length = max(max_length,right - left + 1)
        
        return max_length





            


        